cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+5. A special purpose `Dependencies` class will be introduced to do the above
+mentioned task. This will be responsible for keeping track of dependencies.
+Whenever a bear completes it will be removed from the dependency class.
+6. When all the tasks are done the event loop is stopped and closed.
+7. You would not need to change any bears because coala will handle the
+backwards compatibility. There will be new versions of `LocalBear` and
+`GlobalBear` and will be called `FileWiseBear` and `ProjectWideBear`
+respectively.
+
+# The File Proxy
+
+With the new bear design we are also moving to a new wrapper object to wrap
+every file. This object will also contain some metadata such as the filename
+of the file. The file proxy is an object that contains properties that are
+cached and passed around processes.
+
+What it will store:
+    * The filename, this is the starting point for every other attribute.
+    * The content of the file as a string.
+    * The content of the file, line-splitted
+    * It can also store other things about files as and when required.
+
+A prototype of the file proxy:
+
+```python
+from coala_utils.decorators import generate_eq
+
+
+@generate_eq("filename")
+class FileProxy:
+    def __init__(self, filename):
+        self.filename = filename
+        with open(self.filename) as filehandle:
+            self.content = filehandle.read()
+            self.lines = self.content.splitlines(True)
+
+    def __iter__(self):
+        return iter(self.lines)
+
+    def __hash__(self):
+        return hash(self.filename)
+```
+
+# Dependency Tracking
+
+A new class will be responsible for keeping track of the dependencies of different bear and it will work in the following manner:
+    * The list of all the bears is passed to the class.
+    * It takes this list and runs a method `check_circular_dependency` on it
+    to check if any bears in the list have circular dependencies. If they do
+    then we stop execution.
+    * Each bear will add it's dependencies to this tracker class.
+    * Whenever a bear completes it's execution the `resolve()` method of the
+    new class will be called with this bear as an argument.
+    * This would execute the marking of this bear as complete in the
+    dependency tracker.
+    * Once the marking is complete the tracker returns all bears which have no
+    dependencies remaining as pending, so that their tasks can be scheduled to
+    the process pool.
+
+Here is a prototype:
+
+```python
+from collections import defaultdict
+class CircularDependencyError(Exception):
+
+    @classmethod
+    def for_bears(cls, bears):
+        """
+        Creates the CircularDependencyError with a helpful message about the
+        dependency.
+        """
+        bear_names = [bear.name for bear in bears]
+
+        return cls("Circular dependency detected: " + " -> ".join(bear_names))
```

Comment:
I picked this code up from the current coala code base. So cc @sils1297 if there is any harm in making a constructor out of this?